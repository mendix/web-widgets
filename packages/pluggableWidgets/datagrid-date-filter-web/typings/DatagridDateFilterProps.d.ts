/**
 * This file was generated from DatagridDateFilter.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { CSSProperties } from "react";
import { ActionValue, DynamicValue, EditableValue } from "mendix";

export type DefaultFilterEnum = "between" | "greater" | "greaterEqual" | "equal" | "notEqual" | "smaller" | "smallerEqual" | "empty" | "notEmpty";

export interface DatagridDateFilterContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advanced: boolean;
    defaultValue?: DynamicValue<Date>;
    defaultStartDate?: DynamicValue<Date>;
    defaultEndDate?: DynamicValue<Date>;
    defaultFilter: DefaultFilterEnum;
    placeholder?: DynamicValue<string>;
    adjustable: boolean;
    valueAttribute?: EditableValue<Date>;
    startDateAttribute?: EditableValue<Date>;
    endDateAttribute?: EditableValue<Date>;
    onChange?: ActionValue;
    screenReaderButtonCaption?: DynamicValue<string>;
    screenReaderCalendarCaption?: DynamicValue<string>;
    screenReaderInputCaption?: DynamicValue<string>;
}

export interface DatagridDateFilterPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    advanced: boolean;
    defaultValue: string;
    defaultStartDate: string;
    defaultEndDate: string;
    defaultFilter: DefaultFilterEnum;
    placeholder: string;
    adjustable: boolean;
    valueAttribute: string;
    startDateAttribute: string;
    endDateAttribute: string;
    onChange: {} | null;
    screenReaderButtonCaption: string;
    screenReaderCalendarCaption: string;
    screenReaderInputCaption: string;
}
