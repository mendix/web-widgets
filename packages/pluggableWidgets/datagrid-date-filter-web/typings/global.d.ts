export interface MXLocalePatterns {
    date: string;
    datetime: string;
    time: string;
}

export interface MXSessionLocale {
    code: string;
    firstDayOfWeek: number;
    languageTag: string;
    patterns: MXLocalePatterns;
}

export interface MXSessionConfig {
    locale: MXSessionLocale;
}

export interface MXSession {
    getConfig(): MXSessionConfig;
}

export interface MXGlobalObject {
    session: MXSession;
}

declare global {
    interface Window {
        mx?: MXGlobalObject;
    }
}
