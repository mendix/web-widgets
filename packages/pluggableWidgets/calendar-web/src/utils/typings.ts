import { ObjectItem } from "mendix";
import { CalendarProps as ReactCalendarProps } from "react-big-calendar";
import { withDragAndDropProps } from "react-big-calendar/lib/addons/dragAndDrop";

export interface CalendarEvent {
    title: string;
    start: Date;
    end: Date;
    allDay: boolean;
    color?: string;
    item: ObjectItem;
}

export type EventDropOrResize = {
    event: CalendarEvent;
    start: Date;
    end: Date;
};

export interface DragAndDropCalendarProps<TEvent extends object = Event, TResource extends object = object>
    extends ReactCalendarProps<TEvent, TResource>,
        withDragAndDropProps<TEvent, TResource> {}
