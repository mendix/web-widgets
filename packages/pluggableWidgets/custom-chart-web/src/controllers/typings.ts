import { ActionValue, EditableValue } from "mendix";

export interface ControllerProps {
    configurationOptions: string;
    dataAttribute?: EditableValue<string>;
    dataStatic: string;
    eventDataAttribute?: EditableValue<string>;
    layoutAttribute?: EditableValue<string>;
    layoutStatic: string;
    onClick?: ActionValue;
    sampleData: string;
    sampleLayout: string;
}
