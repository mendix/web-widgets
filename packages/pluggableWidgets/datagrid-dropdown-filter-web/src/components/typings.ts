import { RefFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/RefFilterStore";
import { StaticSelectFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/StaticSelectFilterStore";

export interface EnumFilterAPI {
    filterStore: StaticSelectFilterStore;
    parentChannelName?: string;
}

export interface RefFilterAPI {
    filterStore: RefFilterStore;
    parentChannelName?: string;
}
