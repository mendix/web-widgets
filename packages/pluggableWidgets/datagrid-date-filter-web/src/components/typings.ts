import { Date_InputFilterInterface } from "@mendix/widget-plugin-filtering/typings/InputFilterInterface";

export interface DateFilterProps {
    filterStore: Date_InputFilterInterface;
    parentChannelName?: string;
}
