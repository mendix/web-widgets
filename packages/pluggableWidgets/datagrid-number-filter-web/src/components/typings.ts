import { ChangeEventHandler, CSSProperties } from "react";
import { FilterType } from "../../typings/FilterType";
import { Big } from "big.js";

interface FilterProps {
    adjustable: boolean;
    defaultFilterType: FilterType;
    className?: string;
    id?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    styles?: CSSProperties;
}

export interface FilterInputProps extends FilterProps {
    onFilterTypeClick: (type: FilterType) => void;
    onInputChange: ChangeEventHandler<HTMLInputElement>;
    inputValue: string;
    inputRef?: React.ClassAttributes<HTMLInputElement>["ref"];
    inputDisabled?: boolean;
}

export interface FilterComponentProps extends FilterProps {
    inputChangeDelay: number;
    value?: Big;
    updateFilters?: (value: Big | undefined, type: FilterType) => void;
    name: string;
    datagridChannelName?: string;
}
