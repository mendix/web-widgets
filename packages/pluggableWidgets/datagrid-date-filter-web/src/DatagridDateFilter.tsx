import { createElement, ReactElement } from "react";
import { DatagridDateFilterContainerProps } from "../typings/DatagridDateFilterProps";
import { withAPIv1, withAPIv2 } from "./hocs/withFilterAPI";
import { withPreloader } from "./hocs/withPreloader";
import { FilterComponent } from "./components/FilterComponent";
import { withRuntimeAdapter } from "./hocs/withRuntimeAdapter";

const Component = withAPIv1(withAPIv2(withPreloader(withRuntimeAdapter(FilterComponent))));

export default function DatagridDateFilter(props: DatagridDateFilterContainerProps): ReactElement | null {
    return <Component {...props} />;
}
