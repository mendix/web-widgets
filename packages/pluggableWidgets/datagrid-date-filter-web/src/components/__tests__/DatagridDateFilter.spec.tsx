import "@testing-library/jest-dom";
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { FilterContextValue } from "@mendix/widget-plugin-filtering";
import {
    actionValue,
    dynamicValue,
    EditableValueBuilder,
    ListAttributeValueBuilder
} from "@mendix/widget-plugin-test-utils";
import { mount } from "enzyme";
import { render, fireEvent, screen, act } from "@testing-library/react";
import { createContext, createElement } from "react";
import DatagridDateFilter from "../../DatagridDateFilter";
import { MXGlobalObject, MXSessionConfig } from "../../../typings/global";

function createMXObjectMock(code: string, langTag: string, firstDayOfWeek = 0): MXGlobalObject {
    return {
        session: {
            getConfig: (): MXSessionConfig => ({
                locale: {
                    code,
                    firstDayOfWeek,
                    languageTag: langTag,
                    patterns: {
                        date: "d.M.y",
                        datetime: "d.M.y H.mm",
                        time: "H.mm"
                    }
                }
            })
        }
    };
}

const commonProps = {
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test",
    defaultFilter: "equal" as const,
    adjustable: true,
    advanced: false
};

const mxObject = {
    session: {
        getConfig: () => ({
            locale: {
                languageTag: "en-US",
                patterns: {
                    date: "dd/MM/yyyy"
                }
            }
        })
    }
};

describe("Date Filter", () => {
    describe("with single instance", () => {
        afterEach(() => {
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });

        describe("with single attribute", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("DateTime").withFilterable(true).build()
                } as FilterContextValue);
                (window as any).mx = mxObject;

                jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DatagridDateFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            it("triggers attribute and onchange action on change filter value", () => {
                const action = actionValue();
                const attribute = new EditableValueBuilder<Date>().build();
                render(
                    <DatagridDateFilter
                        {...commonProps}
                        onChange={action}
                        valueAttribute={attribute}
                        placeholder={dynamicValue("Placeholder")}
                    />
                );

                fireEvent.input(screen.getByPlaceholderText("Placeholder"), { target: { value: "01/12/2020" } });

                expect(action.execute).toBeCalledTimes(1);
                expect(attribute.setValue).toBeCalledTimes(1);
            });

            describe("with defaultValue", () => {
                it("initialize with defaultValue", async () => {
                    // 946684800000 = 01.01.2000
                    const date = new Date(946684800000);
                    render(<DatagridDateFilter {...commonProps} defaultValue={dynamicValue<Date>(date)} />);
                    expect(screen.getByRole("textbox")).toHaveValue("01/01/2000");
                });

                it("don't sync value when defaultValue changes from undefined to date", async () => {
                    // 946684800000 = 01.01.2000
                    const date = new Date(946684800000);
                    const { rerender } = render(<DatagridDateFilter {...commonProps} defaultValue={undefined} />);
                    expect(screen.getByRole("textbox")).toHaveValue("");

                    rerender(<DatagridDateFilter {...commonProps} defaultValue={dynamicValue<Date>(date)} />);
                    expect(screen.getByRole("textbox")).toHaveValue("");
                });

                it("don't sync value when defaultValue changes from date to undefined", async () => {
                    // 946684800000 = 01.01.2000
                    const date = new Date(946684800000);
                    const { rerender } = render(
                        <DatagridDateFilter {...commonProps} defaultValue={dynamicValue<Date>(date)} />
                    );
                    expect(screen.getByRole("textbox")).toHaveValue("01/01/2000");

                    rerender(<DatagridDateFilter {...commonProps} defaultValue={undefined} />);
                    expect(screen.getByRole("textbox")).toHaveValue("01/01/2000");
                });
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with multiple attributes", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("DateTime")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("DateTime")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
                (window as any).mx = mxObject;

                jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DatagridDateFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with wrong attribute's type", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("Decimal").withFilterable(true).build()
                } as FilterContextValue);
                (window as any).mx = mxObject;
            });

            it("renders error message", () => {
                const filter = mount(<DatagridDateFilter {...commonProps} />);

                expect(filter.find(Alert).text()).toBe(
                    "The attribute type being used for Date filter is not 'Date and time'"
                );
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with wrong multiple attributes' types", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("String")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("Decimal")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
                (window as any).mx = mxObject;
            });

            it("renders error message", () => {
                const filter = mount(<DatagridDateFilter {...commonProps} />);

                expect(filter.find(Alert).text()).toBe(
                    'The Date filter widget can\'t be used with the filters options you have selected. It requires a "Date and Time" attribute to be selected.'
                );
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with no context", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
                (window as any).mx = mxObject;
            });

            it("renders error message", () => {
                const filter = mount(<DatagridDateFilter {...commonProps} />);

                expect(filter.find(Alert).text()).toBe(
                    "The Date filter widget must be placed inside the header of the Data grid 2.0 or Gallery widget."
                );
            });
        });
    });

    describe("with multiple instances", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                singleAttribute: new ListAttributeValueBuilder().withType("DateTime").withFilterable(true).build()
            } as FilterContextValue);
            (window as any).mx = mxObject;

            jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
        });

        it("renders with a unique id", () => {
            const { asFragment: fragment1 } = render(<DatagridDateFilter {...commonProps} />);
            const { asFragment: fragment2 } = render(<DatagridDateFilter {...commonProps} />);

            expect(fragment1().querySelector("span")?.id).not.toBe(fragment2().querySelector("span")?.id);
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });

    describe("with session config", () => {
        let oldMxObject: MXGlobalObject;
        beforeEach(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                singleAttribute: new ListAttributeValueBuilder().withType("DateTime").withFilterable(true).build()
            } as FilterContextValue);
            oldMxObject = window.mx;
        });

        it("has correct short week days for en-US", async () => {
            window.mx = createMXObjectMock("en_US", "en-US");
            render(<DatagridDateFilter {...commonProps} defaultValue={dynamicValue(new Date("2021-12-10"))} />);

            const open = screen.getByLabelText("Show calendar");
            await act(async () => {
                fireEvent.click(open);
            });
            const header = screen.getByText(/december 2021/gi).parentElement?.lastChild;
            expect(header?.textContent).toEqual("SuMoTuWeThFrSa");
        });

        it("has correct short week days for en-US and starts week at Monday", async () => {
            window.mx = createMXObjectMock("en_US", "en-US", 1);
            render(<DatagridDateFilter {...commonProps} defaultValue={dynamicValue(new Date("2021-12-10"))} />);

            const open = screen.getByLabelText("Show calendar");
            await act(async () => {
                fireEvent.click(open);
            });
            const header = screen.getByText(/december 2021/gi).parentElement?.lastChild;
            expect(header?.textContent).toEqual("MoTuWeThFrSaSu");
        });

        it("has correct short week days for pt-Br", async () => {
            window.mx = createMXObjectMock("pt_BR", "pt-BR");
            render(<DatagridDateFilter {...commonProps} defaultValue={dynamicValue(new Date("2021-12-10"))} />);

            const open = screen.getByLabelText("Show calendar");
            await act(async () => {
                fireEvent.click(open);
            });
            const header = screen.getByText(/dezembro 2021/gi).parentElement?.lastChild;
            expect(header?.textContent).toEqual("domsegterquaquisexsab");
        });

        it("has correct short week days for fi-FI and starts on monday", async () => {
            window.mx = createMXObjectMock("fi_FI", "fi-FI", 1);
            render(<DatagridDateFilter {...commonProps} defaultValue={dynamicValue(new Date("2021-12-10"))} />);

            const open = screen.getByLabelText("Show calendar");
            await act(async () => {
                fireEvent.click(open);
            });
            const header = screen.getByText(/joulukuu 2021/gi).parentElement?.lastChild;
            expect(header?.textContent).toEqual("matiketopelasu");
        });

        it("has correct short week days for fi-FI", async () => {
            window.mx = createMXObjectMock("fi_FI", "fi-FI");
            render(<DatagridDateFilter {...commonProps} defaultValue={dynamicValue(new Date("2021-12-10"))} />);

            const open = screen.getByLabelText("Show calendar");
            await act(async () => {
                fireEvent.click(open);
            });
            const header = screen.getByText(/joulukuu 2021/gi).parentElement?.lastChild;
            expect(header?.textContent).toEqual("sumatiketopela");
        });

        afterEach(() => {
            window.mx = oldMxObject;
        });
    });
});
