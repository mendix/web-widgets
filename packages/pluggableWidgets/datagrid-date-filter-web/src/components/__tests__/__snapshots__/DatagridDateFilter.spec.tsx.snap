// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Filter with single instance with multiple attributes renders correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="filter-selector"
    >
      <div
        class="filter-selector-content"
      >
        <button
          aria-controls="DateFilter1-filter-selectors"
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-default filter-selector-button button-icon equal"
        >
           
        </button>
      </div>
    </div>
    <div
      class="date-filter-container"
      id="datepicker_0.123456789"
    />
    <span
      class="sr-only"
      id="DateFilter1-label"
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container "
      >
        <span
          aria-live="polite"
          class="react-datepicker__aria-live"
          role="alert"
        />
        <input
          aria-labelledby="DateFilter1-label"
          class="form-control filter-input"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      aria-controls="datepicker_0.123456789"
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Show calendar"
      class="btn btn-default btn-calendar"
    >
      <svg
        class="button-icon"
        viewBox="0 0 32 32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Date Filter with single instance with single attribute renders correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="filter-selector"
    >
      <div
        class="filter-selector-content"
      >
        <button
          aria-controls="DateFilter1-filter-selectors"
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-default filter-selector-button button-icon equal"
        >
           
        </button>
      </div>
    </div>
    <div
      class="date-filter-container"
      id="datepicker_0.123456789"
    />
    <span
      class="sr-only"
      id="DateFilter1-label"
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container "
      >
        <span
          aria-live="polite"
          class="react-datepicker__aria-live"
          role="alert"
        />
        <input
          aria-labelledby="DateFilter1-label"
          class="form-control filter-input"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      aria-controls="datepicker_0.123456789"
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Show calendar"
      class="btn btn-default btn-calendar"
    >
      <svg
        class="button-icon"
        viewBox="0 0 32 32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
